### PR TITLE
Kd tree

### LINKED ISSUE #26: Implement Airport Search

### PR DESCRIPTION
**Pull request by galaxyonfire**
closes #26 #25

### CODE REVIEW

**Comment by galaxyonfire** [comment]
unit tests pass on cs1.

**Comment by galaxyonfire** [comment]
You should be able to do a full client -> places -> airport test now. Sadly my airport_server is unable to register :/

File: src/airport_svc.c [hunk 1]
```diff
@@ -11,11 +11,15 @@
...
 #include <sys/socket.h>
 #include <netinet/in.h>
+#include "kdTree.h"
+#include "airport_server.h"
 
 #ifndef SIG_PF
 #define SIG_PF void(*)(int)
 #endif
 
+extern kdTree kd_tree;
```
Comment (QuinnWass) [inline src/airport_svc.c:21]: Consider modeling this off of how I did the prefix trie. However if this works Im fine with it.

File: src/kdTree.h [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+
+#include "airport.h"
+#include <memory>
+#include <queue>
+
+class kdTree
+{
+    public:
+    kdTree() = default;
+    ~kdTree() = default;
```
Comment (QuinnWass) [inline src/kdTree.h:12]: You need to clean up your memory. You are using heap allocated memory.

File: src/kdTree_unit_tests.c [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+    for (size_t i = 0; i < vec_result->size(); ++i)
+        tree.insert(vec_result->at(i));
+
+    airport_info seattle;
+    seattle.latitude = 47.626354;
+    seattle.longitude = -122.333145;
+
+    std::vector<airport_info> list = tree.nearestNeighbors(seattle);
+
+    for (size_t i = 0; i < list.size(); ++i)
```
Comment (QuinnWass) [inline src/kdTree_unit_tests.c:114]: Assert expected outcomes.

**Review by QuinnWass — CHANGES_REQUESTED** [review]

File: src/kdTree.c [hunk 4]
```diff
@@ -19,10 +19,29 @@ const int max_nn = 5;
 
 kdTree::Node::Node(airport_info location) 
 {
-    memcpy ( &(this->location), &location, sizeof (airport_info));
+    this->location = location;
```
Comment (QuinnWass) [inline src/kdTree.c:22]: @Arjunica  Pls confirm this works as expected. I am pretty sure it does.

**Comment by Arjunica** [comment]
I will try but @galaxyonfire needs to merge main in....

**Comment by Arjunica** [comment]
Screen Shot 2020-10-21 at 7 02 54 PM
compiles fine and runs but seems like we getting a negative one return....

**Comment by galaxyonfire** [comment]
@Arjunica there is no -1 case I throw rn, may be a different error

**Review by Arjunica — CHANGES_REQUESTED** [review]
Please merge in main then request review. Thanks =)

File: src/kdTree.h (hunk 2 above)
Comment (galaxyonfire) [reply src/kdTree.h:12]: updated and verified via valgrind

File: src/kdTree_unit_tests.c (hunk 3 above)
Comment (galaxyonfire) [reply src/kdTree_unit_tests.c:208]: added

**Comment by galaxyonfire** [comment]
@Arjunica 
the issue is you never set result.status to 0 on the positive case:
https://github.com/Arjunica/hw4/blob/c8ec45beb1ba0ad754dddf3e869fc6654d5e3b94/src/places_server.c#L76-L93
so places_client never gets result.status == 0

**Comment by galaxyonfire** [comment]
Separate issue but I tried to retype in an answer and I got an infinite loop
image
image

**Comment by Arjunica** [comment]
I'll hotfix it tn

Thanks,
--Arjun

**Comment by Arjunica** [comment]
Lol I got that too. I think it's cuz the error status does don't get
set....

Thanks,
--Arjun

**Comment by galaxyonfire** [comment]
Been trying to enter in correct city/state but keep getting that error :(
image

**Review by Arjunica — APPROVED** [review]

**Pull request merged**